[BREAKING] add matchmissing kwarg to joins

bkamins opened the pull request:
Fixes https://github.com/JuliaData/DataFrames.jl/issues/2499.

This is the core implementation and update of docstrings.

I will add tests later, as it first requires https://github.com/JuliaData/DataFrames.jl/pull/2503 to be merged and rebased against.

nalimilan reviewed (commented):
Looks good. Though I think we should do something about `NaN` (which dplyr also handles via the equivalent of `matchmissing`). Maybe just throw an error if there are `NaN`s for now and see whether people complain.

bkamins commented:
> Maybe just throw an error if there are `NaN`s for now and see whether people complain.

OK. Doing joins on floats is in general not advisable though:
```
julia> outerjoin(DataFrame(x=0.0,a=1), DataFrame(x=-0.0,b=2), on=:x)
2×3 DataFrame
│ Row │ x       │ a       │ b       │
│     │ Float64 │ Int64?  │ Int64?  │
├─────┼─────────┼─────────┼─────────┤
│ 1   │ 0.0     │ 1       │ missing │
│ 2   │ -0.0    │ missing │ 2       │

```
(which means in particular that I am OK to error on `NaN`)

nalimilan commented:
Yeah `-0.0` is also a tricky case. We could also throw an error if we find `-0.0`.

bkamins commented:
@StefanKarpinski - do you have an opinion how `NaN` and `-0.0` should be handled in joins when they are present in columns you join on? (currently we accept them and use `isequal` to match - just like `Dict`s)

As you can see @nalimilan considers if we should not just error on them. In general I understand his point, but I am not 100% sure what would be best. Given your experiences with designing `isequal` in Base what do you think would be best?

kescobo commented:
> Doing joins on floats is in general not advisable though

:100: If people are doing joins on floats, something has already gone seriously wrong. But kudos for trying to get it right. How are other `==` but not `===` things handled? Eg.

```
a = [1,2]
b = [1,2]
c = [1,3]

df1 = DataFrame(x = [a,b,c], y = rand(3))
df2 = DataFrame(x = [a], z = [rand()])
```
```
julia> leftjoin(df1,df2, on=:x)
3×3 DataFrame
│ Row │ x      │ y        │ z        │
│     │ Array… │ Float64  │ Float64? │
├─────┼────────┼──────────┼──────────┤
│ 1   │ [1, 2] │ 0.213345 │ 0.614537 │
│ 2   │ [1, 2] │ 0.70363  │ 0.614537 │
│ 3   │ [1, 3] │ 0.181217 │ missing  │
```

So for consistency, seems like `-0.0` and `0.0` should be treated the same, no?

kescobo commented:
Oh, but the original question was about `NaN` - that one is weird since it's `===` but not `==`... That I have no idea aboutz

tbeason commented:
I think `NaN` could more or less follow whatever decision was made about `missing`, right? Why treat them differently with regards to matching?

I am indifferent about `0.0` and `-0.0` being special cased.

bkamins commented:
I have thought about it and I think I would leave floating point columns as they are now. One normally should not join on such columns anyway, as it is error prone in general. Instead I propose to add a comment in the docstrings warning about doing joins on floating point columns.

nalimilan commented:
Since we can't start throwing an error without breaking, I'd rather throw an error with `NaN`s (current state of PR), and see later whether we should add an argument to disable it (or even disable it altogether). It would be safer to do that also for `-0.0`.

nalimilan commented on src/abstractdataframe/join.jl:
```diff
@@ -41,7 +42,27 @@ struct DataFrameJoiner
+        end
+
+        for df in (dfl_on, dfr_on), col in eachcol(df)
+            if any(x -> (x isa Union{Complex, Real}) & isnan(x), col)
```
Short-circuiting is needed AFAICT:
```suggestion
            if any(x -> (x isa Union{Complex, Real}) && isnan(x), col)
```

bkamins replied:
indeed

bkamins commented:
OK - I will disallow `-0.0` also then and throw an informative error message.

bkamins commented:
I have updated the implementation and added appropriate tests. I think that the solution I propose (to use `CategoricalVector` as a wrapper) is clean enough (ie. if someone really wants to make such a join there is a way to do it).

The PR should be ready for a review.

nalimilan commented on src/abstractdataframe/join.jl:
```diff
@@ -57,8 +57,13 @@ struct DataFrameJoiner
+            if any(x -> (x isa Union{Complex, Real}) &&
+                        (isnan(x) || real(x) === -0.0 || imag(x) === -0.0), col)
+                throw(ArgumentError("currently for numeric values NaN and `-0.0` " *
+                                    "in their real or imaginary components are not" *
+                                    " allowed. Use CategoricalArrays.jl to wrap" *
+                                    "these values into CategoricalVector to perform" *
+                                    "the requested join."))
```
```suggestion
                                    "in their real or imaginary components are not " *
                                    "allowed. Use CategoricalArrays.jl to wrap " *
                                    "these values in a CategoricalVector to perform " *
                                    "the requested join."))
```

nalimilan commented on src/abstractdataframe/join.jl:
```diff
@@ -57,8 +57,13 @@ struct DataFrameJoiner
-            if any(x -> (x isa Union{Complex, Real}) & isnan(x), col)
-                throw(ArgumentError("NaN values in key columns are not allowed"))
+            if any(x -> (x isa Union{Complex, Real}) &&
+                        (isnan(x) || real(x) === -0.0 || imag(x) === -0.0), col)
```
```suggestion
                        (isnan(x) || isequal(real(x), -0.0) || isequal(imag(x), -0.0)), col)
```

bkamins replied:
ah - of course. I was over influenced by your comment not to use `isequal`, but here it is the opposite and it has to be used 😄.

nalimilan commented on test/join.jl:
```diff
@@ -573,25 +573,57 @@ end
+        @test_throws ArgumentError antijoin(name_w_special, job, on=:ID)
+    end
+
+    for special in [missing, 0.0]
```
`0.0` works even without `matchmissing=:equal`, right?

bkamins replied:
right, but I wanted to avoid adding the second set of the same tests. I think it is OK, as `:equal` has no influence on `0.0` handling

bkamins commented:
Thank you!
